File: backend/app/api/v1/problems.py
"""题目接口：列表（分页）+ 详情。"""
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from ...core.errors import ApiError
from ...db import get_db
from ...models.models import Problem
from ...schemas.problem import ProblemDetailOut, ProblemOut
from .deps import get_current_user

router = APIRouter()


@router.get("/problems")
def list_problems(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    keyword: str | None = None,
    db: Session = Depends(get_db),
    user: Problem = Depends(get_current_user),  # noqa: ARG001  仅鉴权
):
    q = db.query(Problem).filter(Problem.defunct.is_(False))
    if keyword:
        q = q.filter(Problem.title.contains(keyword))
    total = q.count()
    items = q.order_by(Problem.id).offset((page - 1) * page_size).limit(page_size).all()
    return {
        "code": 0,
        "message": "ok",
        "data": {
            "total": total,
            "items": [ProblemOut.model_validate(p).model_dump() for p in items],
        },
    }


@router.get("/problems/{pid}")
def problem_detail(
    pid: int,
    db: Session = Depends(get_db),
    user: Problem = Depends(get_current_user),  # noqa: ARG001
):
    p = db.get(Problem, pid)
    if p is None or p.defunct:
        raise ApiError(404, 40401, "problem not found")
    return {"code": 0, "message": "ok", "data": ProblemDetailOut.model_validate(p).model_dump()}

File: frontend/src/api/problems.ts
import http from './http'

export interface ProblemListItem {
  id: number
  title: string
  time_limit: number
  memory_limit: number
  spj: number
}

export interface ProblemDetail extends ProblemListItem {
  description: string
  input: string
  output: string
  sample_input: string
  sample_output: string
  defunct: boolean
}

export interface ProblemListResp {
  total: number
  items: ProblemListItem[]
}

/** GET /api/v1/problems（需登录，分页） */
export async function listProblems(
  page = 1,
  page_size = 20,
  keyword?: string,
): Promise<ProblemListResp> {
  const { data } = await http.get('/problems', { params: { page, page_size, keyword } })
  return data.data
}

/** GET /api/v1/problems/{id}（需登录） */
export async function getProblem(id: number): Promise<ProblemDetail> {
  const { data } = await http.get(`/problems/${id}`)
  return data.data
}
